File: data-structures/queue/queue.ts
type QNode<T> = {
  value: T;
  next?: QNode<T>;
};

export default class Queue<T> {
  public length: number;
  private head?: QNode<T>;
  private tail?: QNode<T>;

  constructor() {
    this.head = this.tail = undefined;
    this.length = 0;
  }

  enqueue(item: T): void {
    const node: QNode<T> = { value: item };

    if (!this.tail) {
      this.head = this.tail = node;
    } else {
      this.tail.next = node;
      this.tail = node;
    }

    this.length++;
  }

  dequeue(): T | undefined {
    if (!this.head) return undefined;

    const removed = this.head;
    this.head = removed.next;
    this.length--;

    if (this.length === 0) {
      this.tail = undefined;
    }

    removed.next = undefined;
    return removed.value;
  }

  peek(): T | undefined {
    return this.head?.value;
  }
}

File: data-structures/queue/queue.py
from typing import Optional, TypeVar, Generic

T = TypeVar('T')

class Node(Generic[T]):
    def __init__(self, value: T):
        self.value = value
        self.next = None

class Queue(Generic[T]):
    def __init__(self):
        self.head: Optional[Node[T]] = None
        self.tail: Optional[Node[T]] = None
        self.length = 0

    def enqueue(self, item: T)->None:
        node = Node(item)
        self.length+=1
        if not self.tail:
            self.tail = self.head = node
            return
        self.tail.next = node
        self.tail = node

    def dequeue(self)->Optional[T]:
        if not self.head:
            return None
        self.length-=1
        head = self.head
        self.head = self.head.next
        return head.value
    
    def peek(self)->Optional[T]:
        return self.head.value if self.head else None

